Remotion/Web/Development.WebTesting/Utilities/ProcessUtils.cs

```diff
@@ -160,14 +206,44 @@ public static void GracefulProcessShutdown ([NotNull] Process process, TimeSpan
         // https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.kill#System_Diagnostics_Process_Kill
         // This exception can be swallowed safely, as we are sure that a process is associated and the process has
         // already exited, therefore the method can return at this point.
-        return;
+      }
+    }
+
+    /// <summary>
+    /// Waits for the specified <paramref name="processes"/> to exit or until the <paramref name="sharedTimeout"/> is reached.
+    /// </summary>
+    /// <param name="processes">The processes that should be waited on to exit.</param>
+    /// <param name="sharedTimeout">The wait timeout after which waiting will be canceled.</param>
+    /// <returns>A subset of the specified <paramref name="processes"/> that did not exit withing the <paramref name="sharedTimeout"/>.</returns>
+    [MustUseReturnValue]
+    private static IReadOnlyList<Process> WaitForProcessesExits (IEnumerable<Process> processes, TimeSpan sharedTimeout)
+    {
+      var remainingProcesses = new List<Process>();
+
+      // Go through all processes and call .WaitForExit on each using a shared timeout
+      // If the timeout runs out still go through all remaining process to make sure to clear any exited processes
+      var stopwatch = Stopwatch.StartNew();
+      foreach (var process in processes)
+      {
+        var timeToTimeoutEnd = sharedTimeout - stopwatch.Elapsed;
+
+        // Use 0 (process.WaitForExit returns instantly) if we exceeded our timeout, otherwise use the remaining timeout
+        var remainingTimeoutInMilliseconds = Math.Max ((int) timeToTimeoutEnd.TotalMilliseconds, 0);
+        if (!process.WaitForExit (remainingTimeoutInMilliseconds))
+          remainingProcesses.Add (process);
       }
 
-      // No exception is thrown when calling .WaitForExit(..) on an already exited process.
-      if (process.WaitForExit (timeoutInMilliseconds))
-        return;
+      return remainingProcesses;
+    }
+
+    private static Exception CreateProcessesDidNotExitInTimeException (IEnumerable<Process> processes, TimeSpan timeout)
+    {
+      var stringBuilder = new StringBuilder();
+      stringBuilder.AppendLine ($"The following processes did not exit within the timeout of {timeout.TotalSeconds:0.##} seconds:");
+      foreach (var process in processes)
```

Comment:
Do you know
`stringBuilder = processes.Aggregate(sb, (p, sb) => sb.AppendLine (..., p.ID))`
Personally, I like the short style, but it depends on how much you like to read linq :D

Also, please use AppendFormat(...).AppendLine(). It's better for memory management.

Reply:
I prefer the "normal" foreach over the `.Aggregate` call in this situation because less work is needed to understand what it does. Aggregate using the same (mutable) variable is effectively a `.ForEach` instead of an `.Aggregate` and makes the code more complicated for not much benefit. Sadly, `.ForEach` is only available on `List<T>` as I would prefer it over a normal foreach iteration.

Reply:
Perhaps. I can work with both and freely admit to having a personal preference to the Aggregate Syntax *). But there is no hard and fast rule :)

*) It goes with the SB acting as a collecting parameter and the Aggregate-pattern, once you know and understand it, just feels more streamlined that way.

Reply:
re-motion rewrite to F# confirmed? 😜

Reply:
We don't talk about Fight Club.